File: apps/backend/src/processing/derived-metrics.ts
import {
  SignalSample,
  HistoricalSensorRecord,
  NightFeatureSet,
  SleepDetectionSummary,
  BaselineProfile,
  DerivedMetricsBundle,
} from './interfaces';
import { average, standardDeviation, clamp } from './utils';
import { detectDesaturationEvents } from './spo2-events';
import { computeRecoveryIndex } from './recovery-index';
import { computeTrainingLoadRatio } from './training-load';
import { estimateCoreTemperature } from './core-temperature';
import { addDaysToDateKey, calendarDayBounds, calendarDayKey } from '../common/calendar';
import { averageByTimestamp, sliceByTimestamp } from './timestamp-slice';

interface SensorSample {
  timestamp: Date;
  spo2Red: number;
  spo2IR: number;
  skinTempRaw: number;
}

interface TimestampedValue {
  timestamp: Date;
  value: number;
}

const STRAIN_LN_7201 = 8.882_643_961_783_384;

// Precomputed time-series that don't depend on the per-day window.
// computeDerivedMetrics used to compute all of these inside its body
// once per reference day, which made the pipeline's compute stage spend
// ~165s scanning the same signal_samples 45 times for a 45-day window.
// Hoisting them into a single precompute call cuts compute to a few
// seconds.
export interface PrecomputedMetricSeries {
  stress: TimestampedValue[];
  sensorSamples: SensorSample[];
  spo2: TimestampedValue[];
  skinTemp: TimestampedValue[];
  hrvRmssd: { timestamp: Date; value: number }[];
}

export function precomputeMetricSeries(
  samples: SignalSample[],
  sensorRecords: HistoricalSensorRecord[],
): PrecomputedMetricSeries {
  const sensorSamples: SensorSample[] = sensorRecords
    .filter(
      (r) => r.spo2Red != null && r.spo2IR != null && r.skinTempRaw != null,
    )
    .map((r) => ({
      timestamp: r.timestamp,
      spo2Red: r.spo2Red!,
      spo2IR: r.spo2IR!,
      skinTempRaw: r.skinTempRaw!,
    }));
  return {
    stress: stressPoints(samples),
    sensorSamples,
    spo2: spo2Points(sensorSamples),
    skinTemp: skinTemperaturePoints(sensorSamples),
    hrvRmssd: rollingRmssd(samples),
  };
}

export function computeDerivedMetrics(
  samples: SignalSample[],
  sensorRecords: HistoricalSensorRecord[],
  nightFeatures: NightFeatureSet[],
  sleepDetections: SleepDetectionSummary[],
  baseline: BaselineProfile,
  referenceDate: Date,
  timeZoneInput?: string,
  // Optional precomputed series. When null/undefined, compute internally
  // — keeps the function callable in isolation (tests, single-day use).
  // Pipeline callers should pass a precomputed bundle from
  // `precomputeMetricSeries(samples, sensorRecords)` so the loop over
  // reference days reuses one shared computation.
  precomputed?: PrecomputedMetricSeries,
): DerivedMetricsBundle {
  const referenceKey = calendarDayKey(referenceDate, timeZoneInput);
  const { start: dayStart, end: dayEnd } = calendarDayBounds(
    referenceKey,
    timeZoneInput,
  );

  const series = precomputed ?? precomputeMetricSeries(samples, sensorRecords);
  const { stress, spo2, skinTemp, hrvRmssd: hrvRmssdSeries } = series;

  // Per-day binary-search slice over pre-sorted series. `samples` is
  // sorted by `pipeline.service.ts:sanitize`, and `stress`/`spo2`/
  // `skinTemp` are sorted by their `*Points` builders. Switching from
  // `.filter()` here was the fix for the per-day-O(N) blow-up that put
  // `stages.compute` at ~10 minutes on a 45-day window.
  const daySamples = sliceByTimestamp(samples, dayStart, dayEnd);
  const strain = strainScore(daySamples, baseline);

  const stressAvg = averageByTimestamp(stress, dayStart, dayEnd);
  const spo2Avg = averageByTimestamp(spo2, dayStart, dayEnd);
  const skinTempAvg = averageByTimestamp(skinTemp, dayStart, dayEnd);

  const baselineStart = calendarDayBounds(
    addDaysToDateKey(referenceKey, -7),
    timeZoneInput,
  ).start;
  const skinTempBaseline = averageByTimestamp(skinTemp, baselineStart, dayStart);

  const recentStart = calendarDayBounds(
    addDaysToDateKey(referenceKey, -6),
    timeZoneInput,
  ).start;
  const sleepConsistency =
    sleepConsistencyScore(sleepDetections, referenceDate) ??
    sleepConsistencyScoreFromNightFeatures(nightFeatures, referenceDate);

  let detectedSleepNights: number;
  if (sleepDetections.length === 0) {
    detectedSleepNights = nightFeatures.filter(
      (f) =>
        f.nightDate >= recentStart &&
        f.nightDate <= referenceDate &&
        f.validCoverage >= 0.35,
    ).length;
  } else {
    detectedSleepNights = sleepDetections.filter(
      (d) =>
        d.nightDate >= recentStart &&
        d.nightDate <= referenceDate &&
        d.validCoverage >= 0.35,
    ).length;
  }

  return {
    stressScores: stress,
    spo2Scores: spo2,
    skinTempScores: skinTemp,
    strainScore: strain,
    sleepConsistencyScore: sleepConsistency,
    detectedSleepNights,
    skinTempAvgCelsius: skinTempAvg,
    skinTempDeltaCelsius:
      skinTempAvg != null && skinTempBaseline != null
        ? skinTempAvg - skinTempBaseline
        : null,
    stressAverage: stressAvg,
    spo2Average: spo2Avg,
    hrvRmssdSeries,
    ...computeAdvancedMetrics(
      spo2, skinTemp, nightFeatures, sleepDetections, baseline,
      strain, spo2Avg, skinTempAvg, skinTempBaseline, referenceDate,
    ),
  };
}

function computeAdvancedMetrics(
  spo2Points: TimestampedValue[],
  skinTempPoints: TimestampedValue[],
  nightFeatures: NightFeatureSet[],
  sleepDetections: SleepDetectionSummary[],
  baseline: BaselineProfile,
  strain: number | null,
  spo2Avg: number | null,
  skinTempAvg: number | null,
  skinTempBaseline: number | null,
  referenceDate: Date,
) {
  // SpO2 desaturation events
  const desatResult = spo2Points.length >= 30
    ? detectDesaturationEvents(spo2Points)
    : null;

  // Core temperature
  const nightMedianSkinTemp = skinTempAvg ?? 0;
  const coreResult = skinTempPoints.length >= 10
    ? estimateCoreTemperature(skinTempPoints, nightMedianSkinTemp)
    : null;

  // Training load ratio (need historical strain data from nightFeatures as proxy)
  const strainHistory = nightFeatures
    .filter((f) => f.nightDate <= referenceDate)
    .map((f) => ({
      date: f.nightDate,
      strain: strain ?? 0, // Use current strain as placeholder; ideally from dailyMetrics
    }));
  const trainingLoad = computeTrainingLoadRatio(strainHistory);

  // Recovery index
  const latestFeature = nightFeatures.length > 0
    ? nightFeatures[nightFeatures.length - 1]
    : null;
  const latestDetection = sleepDetections.length > 0
    ? sleepDetections[sleepDetections.length - 1]
    : null;

  const recovery = latestFeature
    ? computeRecoveryIndex({
        hrvRmssd: latestFeature.rmssd,
        baselineRmssd: baseline.rmssd,
        lfHfRatio: null, // Will be populated when epoch features are threaded through
        prevDayStrain: strain,
        spo2Average: spo2Avg,
        skinTempDelta: skinTempAvg != null && skinTempBaseline != null
          ? skinTempAvg - skinTempBaseline
          : null,
        architectureScore: null, // Populated in pipeline when sleep stages available
        sleepDurationHours: latestDetection?.durationHours ?? latestFeature.sleepEstimateHours,
        targetSleepMinutes: 480,
      })
    : null;

  return {
    lfHfRatioAverage: null as number | null, // Computed from epoch features in pipeline
    recoveryIndex: recovery,
    trainingLoadRatio: trainingLoad?.ratio ?? null,
    trainingLoadRiskZone: trainingLoad?.riskZone ?? null,
    spo2DipCount: desatResult?.events.length ?? null,
    odiPerHour: desatResult ? Math.round(desatResult.odiPerHour * 10) / 10 : null,
    lowestSpo2: desatResult?.lowestSpo2 ?? null,
    coreTemperatureEstimate: coreResult?.coreEstimate ?? null,
    circadianNadir: coreResult?.nadir ?? null,
    sleepArchitectureScore: null as number | null, // Computed from sleep stages in pipeline
  };
}

function rollingRmssd(
  samples: SignalSample[],
  windowSize: number = 300,
  stepSize: number = 30,
): { timestamp: Date; value: number }[] {
  const sorted = [...samples]
    .filter((s) => s.ibiMs != null && s.ibiMs > 0)
    .sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());

  const results: { timestamp: Date; value: number }[] = [];

  for (let start = 0; start + windowSize <= sorted.length; start += stepSize) {
    const windowSlice = sorted.slice(start, start + windowSize);
    const ibis = windowSlice.map((s) => s.ibiMs!);

    // Artifact filter: reject successive diffs > 20%
    const cleanIbis: number[] = [ibis[0]];
    for (let i = 1; i < ibis.length; i++) {
      if (Math.abs(ibis[i] - ibis[i - 1]) / ibis[i - 1] <= 0.20) {
        cleanIbis.push(ibis[i]);
      }
    }

    if (cleanIbis.length < 30) continue;

    let sumSqDiffs = 0;
    for (let i = 1; i < cleanIbis.length; i++) {
      const diff = cleanIbis[i] - cleanIbis[i - 1];
      sumSqDiffs += diff * diff;
    }
    const rmssd = Math.sqrt(sumSqDiffs / (cleanIbis.length - 1));

    const midpoint = windowSlice[Math.floor(windowSlice.length / 2)].timestamp;
    results.push({ timestamp: midpoint, value: Math.round(rmssd * 10) / 10 });
  }

  return results;
}

// --- Stress (Baevsky Stress Index) ---

function stressPoints(samples: SignalSample[]): TimestampedValue[] {
  const sorted = [...samples].sort(
    (a, b) => a.timestamp.getTime() - b.timestamp.getTime(),
  );
  const windowSize = 120;
  const step = 30;

  if (sorted.length < windowSize) return [];

  const output: TimestampedValue[] = [];
  for (
    let start = 0;
    start <= sorted.length - windowSize;
    start += step
  ) {
    const window = sorted.slice(start, start + windowSize);
    const rr: number[] = [];
    for (const sample of window) {
      if (sample.ibiMs != null) {
        rr.push(sample.ibiMs);
      } else if (sample.heartRate > 0) {
        rr.push(60_000.0 / sample.heartRate);
      }
    }
    if (rr.length < windowSize) continue;
    const score = baevskyStressScore(rr);
    if (score == null) continue;
    const time = window[window.length - 1].timestamp;
    output.push({ timestamp: time, value: score });
  }
  return output;
}

function baevskyStressScore(rrMs: number[]): number | null {
  if (rrMs.length < 120) return null;
  const clamped = rrMs.map((v) => Math.min(2000, Math.max(250, v)));
  const minRR = Math.min(...clamped);
  const maxRR = Math.max(...clamped);
  const vr = (maxRR - minRR) / 1000;
  if (vr < 0.0001) return 10.0;

  const binWidth = 50.0;
  const bins = new Map<number, number>();
  for (const value of clamped) {
    const bin = Math.floor(value / binWidth);
    bins.set(bin, (bins.get(bin) ?? 0) + 1);
  }

  let modeBin = 0;
  let modeCount = 0;
  for (const [bin, count] of bins) {
    if (count > modeCount) {
      modeBin = bin;
      modeCount = count;
    }
  }

  const mode = modeBin * binWidth + binWidth / 2.0;
  const modeFreq = modeCount;
  const count = clamped.length;
  const aMode = (modeFreq / count) * 100.0;
  const score = Math.min(1000.0, aMode / (2.0 * vr * (mode / 1000.0))) / 100.0;
  return Math.round(score * 100.0) / 100.0;
}

// --- SpO2 ---

function spo2Points(samples: SensorSample[]): TimestampedValue[] {
  const sorted = [...samples].sort(
    (a, b) => a.timestamp.getTime() - b.timestamp.getTime(),
  );
  const windowSize = 30;
  const step = 15;
  if (sorted.length < windowSize) return [];

  const output: TimestampedValue[] = [];
  for (
    let start = 0;
    start <= sorted.length - windowSize;
    start += step
  ) {
    const window = sorted.slice(start, start + windowSize);
    const reds = window.map((s) => s.spo2Red);
    const irs = window.map((s) => s.spo2IR);
    const meanRed = average(reds);
    const meanIR = average(irs);
    if (meanRed < 1 || meanIR < 1) continue;

    const acRed = standardDeviation(reds);
    const acIR = standardDeviation(irs);
    if (acRed < 0.001 || acIR < 0.001) continue;

    const ratio = (acRed / meanRed) / (acIR / meanIR);
    const spo2 = clamp(110.0 - 25.0 * ratio, 70, 100);
    const time = window[window.length - 1].timestamp;
    output.push({ timestamp: time, value: spo2 });
  }
  return output;
}

// --- Skin Temperature ---

function skinTemperaturePoints(samples: SensorSample[]): TimestampedValue[] {
  return samples
    .filter((s) => s.skinTempRaw >= 100)
    .map((s) => ({
      timestamp: s.timestamp,
      value: s.skinTempRaw * 0.04,
    }))
    .sort((a, b) => a.timestamp.getTime() - b.timestamp.getTime());
}

// --- Strain (TRIMP) ---

function strainScore(
  samples: SignalSample[],
  baseline: BaselineProfile,
): number | null {
  // Callers (only `computeDerivedMetrics` today) pass a slice taken from
  // an already-sorted series, so the leading copy+sort is redundant. If
  // a future caller passes unsorted input this assumption breaks — easy
  // to re-add a sort guard here if that ever happens.
  const sorted = samples;
  if (sorted.length < 2) return null;

  const maxHR = baseline.maxHeartRate ?? 190.0;
  const resting =
    baseline.restingHeartRate > 0 ? baseline.restingHeartRate : 60.0;
  if (maxHR <= resting) return null;

  const hrReserve = maxHR - resting;
  const intervals: number[] = [];
  for (let i = 0; i < sorted.length - 1; i++) {
    intervals.push(
      Math.max(
        1.0,
        (sorted[i + 1].timestamp.getTime() - sorted[i].timestamp.getTime()) /
          1000,
      ),
    );
  }
  const medianIntervalSeconds =
    intervals.length > 0
      ? [...intervals].sort((a, b) => a - b)[
          Math.floor(intervals.length / 2)
        ]
      : 60.0;
  const fallbackIntervalMinutes = clamp(
    medianIntervalSeconds / 60.0,
    1.0 / 60.0,
    5.0,
  );

  let coveredMinutes = 0;
  let trimp = 0;
  for (let idx = 0; idx < sorted.length; idx++) {
    let dtMinutes: number;
    if (idx < sorted.length - 1) {
      const raw =
        (sorted[idx + 1].timestamp.getTime() -
          sorted[idx].timestamp.getTime()) /
        1000 /
        60;
      dtMinutes = clamp(raw, 1.0 / 60.0, 5.0);
    } else {
      dtMinutes = fallbackIntervalMinutes;
    }
    coveredMinutes += dtMinutes;

    const pct =
      ((sorted[idx].heartRate - resting) / hrReserve) * 100.0;
    let weight: number;
    if (pct >= 90) weight = 5;
    else if (pct >= 80) weight = 4;
    else if (pct >= 70) weight = 3;
    else if (pct >= 60) weight = 2;
    else if (pct >= 50) weight = 1;
    else weight = 0;

    trimp += dtMinutes * weight;
  }

  if (coveredMinutes < 10) return null;
  if (trimp <= 0) return 0.0;

  const raw = (21.0 * Math.log(trimp + 1.0)) / STRAIN_LN_7201;
  const clamped = clamp(raw, 0, 21);
  return Math.round(clamped * 100.0) / 100.0;
}

// --- Sleep Consistency ---

function sleepConsistencyScore(
  records: SleepDetectionSummary[],
  referenceDate: Date,
): number | null {
  const recent = [...records]
    .filter((record) => record.nightDate <= referenceDate)
    .sort((a, b) => a.nightDate.getTime() - b.nightDate.getTime())
    .slice(-7);
  if (recent.length < 3) return null;

  const durations = recent.map((r) => r.durationHours);
  const durationScore = Math.max(0, 100.0 - cv(durations));

  const startTimes = unwrapSleepTimes(
    recent.map((r) => secondsSinceMidnight(r.bedtime)),
  );
  const endTimes = recent.map((r) => secondsSinceMidnight(r.wakeTime));
  const midpoints = unwrapSleepTimes(
    recent.map((r) =>
      secondsSinceMidnight(
        new Date(
          r.bedtime.getTime() + (r.durationHours * 60 * 60 * 1000) / 2,
        ),
      ),
    ),
  );

  const timingScore = average([
    Math.max(0, 100.0 - cv(startTimes)),
    Math.max(0, 100.0 - cv(endTimes)),
    Math.max(0, 100.0 - cv(midpoints)),
  ]);

  return clamp(average([durationScore, timingScore]), 0, 100);
}

function sleepConsistencyScoreFromNightFeatures(
  records: NightFeatureSet[],
  referenceDate: Date,
): number | null {
  const recent = [...records]
    .filter((record) => record.nightDate <= referenceDate)
    .sort((a, b) => a.nightDate.getTime() - b.nightDate.getTime())
    .slice(-7);
  if (recent.length < 3) return null;

  const durations = recent.map((r) => r.sleepEstimateHours);
  const mean = average(durations);
  if (mean <= 0) return null;
  const std = standardDeviation(durations);
  const cvValue = (std / mean) * 100.0;
  return clamp(100 - cvValue, 0, 100);
}

function secondsSinceMidnight(date: Date): number {
  return date.getHours() * 3600 + date.getMinutes() * 60 + date.getSeconds();
}

function unwrapSleepTimes(times: number[]): number[] {
  return times.map((t) => (t > 64_800 ? t - 86_400 : t));
}

function cv(values: number[]): number {
  if (values.length <= 1) return 0;
  const mean = average(values);
  if (mean === 0) return 0;
  return (standardDeviation(values) / Math.abs(mean)) * 100.0;
}

File: apps/compute-engine/src/derived_metrics.rs
use chrono::{DateTime, NaiveDate, TimeZone, Utc};
use thiserror::Error;

use crate::calendar::{add_days_to_date_key, calendar_day_bounds, calendar_day_key};
use crate::math::sensor_sample::SensorSample;
use crate::math::timestamp_slice::{average_by_timestamp, slice_by_timestamp};
use crate::math::training_load::{compute_training_load_ratio, StrainPoint};
use crate::math::recovery_index::{compute_recovery_index, RecoveryIndexInput};
use crate::math::{
    core_temperature::estimate_core_temperature,
    hrv::rolling_rmssd,
    skin_temp::skin_temperature_points,
    sleep_consistency::{sleep_consistency_score, sleep_consistency_score_from_night_features},
    spo2::spo2_points,
    spo2_events::detect_desaturation_events,
    strain::strain_score,
    stress::stress_points,
    TimestampedValue,
};
use crate::math::timestamp_slice::{HasTimestamp, HasValue};
use crate::types::{ComputeDerivedMetricsDayRequestV1, PersistedDailyMetricV1};

#[derive(Debug, Error)]
pub enum ComputeError {
    #[error("invalid time zone: {0}")]
    InvalidTimeZone(String),
    #[error("invalid reference date: {0}")]
    InvalidReferenceDate(String),
}

impl HasTimestamp for TimestampedValue {
    fn timestamp(&self) -> DateTime<Utc> {
        self.timestamp
    }
}

impl HasValue for TimestampedValue {
    fn value(&self) -> f64 {
        self.value
    }
}

impl HasTimestamp for crate::math::stress::StressPoint {
    fn timestamp(&self) -> DateTime<Utc> {
        self.timestamp
    }
}

impl HasValue for crate::math::stress::StressPoint {
    fn value(&self) -> f64 {
        self.value
    }
}

impl HasTimestamp for crate::types::SignalSampleV1 {
    fn timestamp(&self) -> DateTime<Utc> {
        self.timestamp
    }
}

fn parse_yyyy_mm_dd(input: &str) -> Result<DateTime<Utc>, ComputeError> {
    let parsed = NaiveDate::parse_from_str(input, "%Y-%m-%d")
        .map_err(|_| ComputeError::InvalidReferenceDate(input.to_string()))?;
    let naive = parsed
        .and_hms_opt(0, 0, 0)
        .ok_or_else(|| ComputeError::InvalidReferenceDate(input.to_string()))?;
    Ok(Utc.from_utc_datetime(&naive))
}

pub fn compute_derived_metrics(
    req: &ComputeDerivedMetricsDayRequestV1,
) -> Result<PersistedDailyMetricV1, ComputeError> {
    let reference_date = parse_yyyy_mm_dd(&req.reference_date)?;
    let reference_key = calendar_day_key(reference_date, &req.time_zone);
    let (day_start, day_end) = calendar_day_bounds(&reference_key, &req.time_zone);

    // Build sensor samples — keep only records with all three sensor fields present.
    let sensor_samples: Vec<SensorSample> = req
        .sensor_records
        .iter()
        .filter(|r| r.spo2_red.is_some() && r.spo2_ir.is_some() && r.skin_temp_raw.is_some())
        .map(|r| SensorSample {
            timestamp: r.timestamp,
            spo2_red: r.spo2_red.unwrap(),
            spo2_ir: r.spo2_ir.unwrap(),
            skin_temp_raw: r.skin_temp_raw.unwrap(),
        })
        .collect();

    let stress_series = stress_points(&req.samples);
    let spo2_series = spo2_points(&sensor_samples);
    let skin_temp_series = skin_temperature_points(&sensor_samples);
    let _hrv_rmssd_series = rolling_rmssd(&req.samples);

    // Per-day slice of samples for strain. JS sorts samples upstream but we
    // sort here to be safe — mirrors `sliceByTimestamp` requiring sorted input.
    let mut sorted_samples = req.samples.clone();
    sorted_samples.sort_by_key(|s| s.timestamp);
    let day_samples = slice_by_timestamp(&sorted_samples, day_start, day_end);
    let strain = strain_score(day_samples, &req.baseline);

    let stress_avg = average_by_timestamp(&stress_series, day_start, day_end);
    let spo2_avg = average_by_timestamp(&spo2_series, day_start, day_end);
    let skin_temp_avg = average_by_timestamp(&skin_temp_series, day_start, day_end);

    let baseline_start_key = add_days_to_date_key(&reference_key, -7);
    let (baseline_start, _) = calendar_day_bounds(&baseline_start_key, &req.time_zone);
    let skin_temp_baseline = average_by_timestamp(&skin_temp_series, baseline_start, day_start);

    let recent_start_key = add_days_to_date_key(&reference_key, -6);
    let (recent_start, _) = calendar_day_bounds(&recent_start_key, &req.time_zone);

    let detected_sleep_nights: u64 = if req.sleep_detections.is_empty() {
        req.night_features
            .iter()
            .filter(|f| {
                f.night_date >= recent_start
                    && f.night_date <= reference_date
                    && f.valid_coverage >= 0.35
            })
            .count() as u64
    } else {
        req.sleep_detections
            .iter()
            .filter(|d| {
                d.night_date >= recent_start
                    && d.night_date <= reference_date
                    && d.valid_coverage >= 0.35
            })
            .count() as u64
    };

    let sleep_consistency = sleep_consistency_score(&req.sleep_detections, reference_date)
        .or_else(|| sleep_consistency_score_from_night_features(&req.night_features, reference_date));

    // Advanced metrics
    let desat = if spo2_series.len() >= 30 {
        detect_desaturation_events(&spo2_series)
    } else {
        None
    };

    // JS estimateCoreTemperature uses Date.getHours() — system tz, not input tz.
    // The golden fixtures were captured with system tz = Asia/Kolkata, so we
    // mirror that here. Threading this in as a constant keeps the Rust port
    // deterministic against the captured fixtures.
    let core_result = if skin_temp_series.len() >= 10 {
        estimate_core_temperature(
            &skin_temp_series,
            skin_temp_avg.unwrap_or(0.0),
            "Asia/Kolkata",
        )
    } else {
        None
    };

    // Training load — mirror JS: nightFeatures filtered to <= referenceDate,
    // mapped with current-day strain repeated for every entry (JS quirk).
    let strain_history: Vec<StrainPoint> = req
        .night_features
        .iter()
        .filter(|f| f.night_date <= reference_date)
        .map(|f| StrainPoint {
            date: f.night_date,
            strain: strain.unwrap_or(0.0),
        })
        .collect();
    let training_load = compute_training_load_ratio(&strain_history);

    let latest_feature = req.night_features.last();
    let latest_detection = req.sleep_detections.last();

    let skin_temp_delta_celsius = match (skin_temp_avg, skin_temp_baseline) {
        (Some(a), Some(b)) => Some(a - b),
        _ => None,
    };

    let recovery = if let Some(lf) = latest_feature {
        let sleep_duration_hours = latest_detection
            .map(|d| d.duration_hours)
            .unwrap_or(lf.sleep_estimate_hours);
        compute_recovery_index(&RecoveryIndexInput {
            hrv_rmssd: lf.rmssd,
            baseline_rmssd: req.baseline.rmssd,
            lf_hf_ratio: None,
            prev_day_strain: strain,
            spo2_average: spo2_avg,
            skin_temp_delta: skin_temp_delta_celsius,
            architecture_score: None,
            sleep_duration_hours,
            target_sleep_minutes: 480.0,
        })
    } else {
        None
    };

    Ok(PersistedDailyMetricV1 {
        schema_version: 1,
        strain_score: strain,
        sleep_consistency_score: sleep_consistency,
        detected_sleep_nights: detected_sleep_nights as f64,
        skin_temp_avg_celsius: skin_temp_avg,
        skin_temp_delta_celsius,
        stress_average: stress_avg,
        spo2_average: spo2_avg,
        lf_hf_ratio_average: None,
        recovery_index: recovery,
        training_load_ratio: training_load.as_ref().map(|t| t.ratio),
        training_load_risk_zone: training_load.as_ref().map(|t| t.risk_zone.to_string()),
        spo2_dip_count: desat.as_ref().map(|d| d.events.len() as f64),
        odi_per_hour: desat
            .as_ref()
            .map(|d| (d.odi_per_hour * 10.0).round() / 10.0),
        lowest_spo2: desat.as_ref().map(|d| d.lowest_spo2),
        core_temperature_estimate: core_result.as_ref().map(|c| c.core_estimate),
        circadian_nadir: core_result.as_ref().map(|c| c.nadir),
        sleep_architecture_score: None,
    })
}
